docs/conventions/index.rst

```diff
@@ -0,0 +1,141 @@
+===========
+Conventions
+===========
+
+Why Conventions?
+~~~~~~~~~~~~~~~~
+
+Zarr Conventions provide a mechanism to standardize metadata and layout of Zarr data
+in order to meet domain-specific application needs without changes to the
+core Zarr data model and specification, and without specification extensions.
+
+Conventions must fit completely within the Zarr data / metadata model of groups, arrays, and attributes thereof, requiring
+no changes or extension to the specification.
+A Zarr implementation itself should not even be aware of the existence of the convention.
+The line between a convention and an extension may be blurry in some cases.
+The key distinction lies in the implementation: the responsibility for interpreting a *convention* relies completely with downstream,
+domain-specific software, while an *extension* must be handled by the Zarr implementation itself.
+A good rule of thumb is that a user should be able to safely ignore the convention and still be able to interact with the data via the core Zarr library,
+even if some domain-specific context or functionality is missing.
+If the data are completely meaningless or unintelligible without the convention, then it should be an extension instead.
+
+Conventions can also help users switch between different storage libraries more flexibly.
+Since Zarr and HDF5 implement nearly identical data models, a single convention can be applied to both formats.
+This allows downstream software to maintain better separation of concerns between storage and domain-specific logic.
+
+Conventions are modular and composable. A single group or array can conform to multiple conventions.
+
+
+Describing Conventions
+~~~~~~~~~~~~~~~~~~~~~~
+
+Conventions Document
+--------------------
+
+Conventions are described by a *convention document*.
+TODO: say more about the structure and format of this document
+
+Explicit Conventions
+--------------------
+
+The preferred way of identifying the presence of a convention in a Zarr group or array is via the attribute `zarr_conventions`.
+This attribute must be an array of strings; each string is an identifier for the convention.
+Multiple conventions may be present.
+
+For example, a group metadata JSON document with conventions present might look like this
+
+.. code-block:: json
+
+   {
+      "zarr_format": 3,
+      "node_type": "group",
+      "attributes": {
+         "zarr_conventions": ["units-v1", "foo"],
```

Comment:
any chance to make it more "specific" but also descriptive to potentially "decentralize" such conventions, while still allowing for a generic validation of zarrs. E.g. it could become here a dict of conversions, with their versions and schema (jsonschema ? or may be linkml?) URLs . e.g.

```suggestion
         "zarr_conventions": {
           "units": {
              "version": 1, 
              "homepage": " ... URL which has potential to describe what that is about ...",
              "schema-url": "... hosted somewhere ..."
            },
          "foo":  {}
          },
```

where in above `units` is a well defined convention and `foo` is not so good (just for an example).

Providing schema to go along would open opportunity for a generic zarr validator to validate embedded in a zarr attributes following the schema. It is reflective of an approach NWB standard took - it stores a copy of the schema for itself of each of the extensions within .nwb (hdf5) file so it becomes feasible to do generic validation and also open it up following those embedded schemas even if extension library is not installed. 

Separation of version from the convention name also would make it cleaner and `diff` upon upgrade from one version to another becoming "to the point" (instead of changing every attribute name) thus making it easier to review etc.

I am not that savvy in zarr and thus acknowledge that development of the schema formalization for conventions might be a larger effort than intended for this ZEP, so might better be postponed. But establishing record of `zarr_conventions` as a collection of records instead of just a list, would at least open such possibility without in the future requiring breaking type changes. Or may be it is already "easy" to add basic "schema" support here?

Reply:
@yarikoptic should we rename `schema-url` to `schemaUrl` to adhere to JSON common practices? Hyphens, when parsed in some languages, cause issues.

Reply:
What exactly is the use case for storing a schema for the attributes alongside the attributes? I don't see how validation is attractive in this situation, because presumably if the attributes don't pass validation from that schema, you wouldn't write them to disk in the first place. I'm a client reading a Zarr group that implements some schema that I am aware of, then by definition I already have the schema, so including the schema in the Zarr attributes is useless here; whereas, if it's a schema I'm _not_ aware of, then why should I care if validation of that schema succeeds or fails?

I can see why a data stores that support partial reads would expose schemas, because you don't want clients to read everything just to know what's in it, but Zarr attributes are just JSON documents, so partial reading isn't really part of the picture there.

It's very likely that I don't understand the use case, so a motivating example would really help here.

Reply:
> I don't see how validation is attractive in this situation, because presumably if the attributes don't pass validation from that schema, you wouldn't write them to disk in the first place.

That is quite a big assumption which would be impossible to verify unless schema is stored/pointed to explicitly. "Explicit better than implicit" (Zen of Python `#2`).  There can be a number of buggy client implementations, etc. Absent formalization of schema on Zarr level would facilitate "schema-free" *conventions* down-stream, thus facilitate  breeding unformalized conventions/extensions.

Besides validation, having a schema over the fields might open opportunities for automated metadata-visualization/editing UI constructions (e.g. using smth like https://github.com/koumoul-dev/vuetify-jsonschema-form/ for vue) etc. 

FWIW, having machine readable schema is a great feature for a standard to have: e.g. a foundational design principle within https://www.nwb.org/ (https://github.com/NeurodataWithoutBorders/nwb-schema),  and recently (well -- years back but still being formalized) established within https://bids-specification.readthedocs.io/ ([src/schema](https://github.com/bids-standard/bids-specification/tree/master/src/schema)), but already acknowledged to be of great importance.

Reply:
> @yarikoptic should we rename `schema-url` to `schemaUrl` to adhere to JSON common practices? Hyphens, when parsed in some languages, cause issues / require special handling.

good point!  I replaced with `schema_url` though to match "convention" among other fields in this PR.  Is there a generic recommendation for either `_`-joined or `camelCase` in zarr metadata attributes?

Reply:
I'm 100% team schema -- I use schemas extensively for Zarr data. So I'm not asking "why are schemas good", I'm asking why the data being schematized needs to also know the address of its schema.

Reply:
> good point! I replaced with `schema_url` though to match "convention" among other fields in this PR. Is there a generic recommendation for either `_`-joined or `camelCase` in zarr metadata attributes?

the core zarr metadata uses `snake_case`, so `schema_url` fits!

Reply:
@d-v-b one concrete use case for me is:

We have multiple business units that access Zarr files and sometimes they're in strictly typed languages. We have very detailed metadata (like units, statistics) that define nested hierarchies of floats/ints etc.

By providing schema, parsing and validating this data to the correct structure in the language of choice becomes trivial.

It also stops people from writing files that are not x-compatible.

Reply:
> By providing schema, parsing and validating this data to the correct structure in the language of choice becomes trivial.
>
> It also stops people from writing files that are not x-compatible.

I'm 100% with you on these benefits of having a schema. As I said upthread, I support and use schemas for zarr hierarchies, for the reasons you describe. What I'm asking is: what value do we get from the `schema_url` field, and more broadly, enumerating the schemas explicitly?

Reply:
here's a concrete example of what I am worried about: suppose there are two conventions for Zarr array attributes: A and B (maybe A is units, and B is rendering settings). Now suppose a community starts using Zarr arrays with conventions A and B often enough that "A _and_ B" becomes its own convention, which they call C. If we are listing conventions by name in zarr attributes, it's ambiguous whether we would use `"zarr_conventions": ["convention_A", "convention_B"]` or `"zarr_conventions": ["convention_C"]`. This is not a fake example -- I maintain a [schematized zarr hierarchy definition](https://janelia-cellmap.github.io/cellmap-schemas/api/annotation/#cellmap_schemas.annotation.CropGroup) that is an extension of OME-NGFF, which is itself a zarr hierarchy convention.

Reply:
without either explicit URL to the schema (typical approach in linked-web, e.g. .jsonld serializations of RDF), or embedding entire schema inside (done e.g. by nwb) you would need some centralized registry of those per name space.  Such centralization might introduce bottleneck and actually complicate adoption of such method.

NB in general embedding schemas to .zarr itself might be a good/robust solution.. in principle I guess `schema_url` could be generalized to `schema_uri` and support smth like `file:/./schemas/convention.json`? (this is just a wild idea, not even suere if LGTM URI here... but URLs AFAIK unfortunately would not work since should be absolute)

Reply:
> without either explicit URL to the schema (typical approach in linked-web, e.g. .jsonld serializations of RDF), or embedding entire schema inside (done e.g. by nwb) you would need some centralized registry of those per name space.

I would suggest addressing this by simply not requiring the Zarr array / group name the conventions it adheres to. Typing would be structural, and there is no name to find in a namespace.

Reply:
not exactly sure what you propose to reply constructively, but sounds like "dump everything into a dict" which would be counter-effective to the original intention of this ZEP to  (citing from https://zarr.dev/zeps/draft/ZEP0004.html; emphasis is mine)

> .. **standardize** conventions around metadata and layout of Zarr data using user-defined attributes ...

Reply:
It is correct that I don't agree with the proposal of the ZEP, insofar as it proposes to embed schema / type information inside the thing being schematized / typed, but I'm not really advocating for "dumping everything in a dict" either. 

What I advocate is very simple: Schemas and associated tooling should be used to generate and validate zarr hierarchies. E.g., defining Zarr hierarchies as typed data, and checking that instances of Zarr hierarchies pass type-checking. See [pydantic-zarr](https://janelia-cellmap.github.io/pydantic-zarr/usage_zarr_v2/#using-generic-types) for an example of this approach. This is a very simple idea: take some unstructured data, apply a type system to it, get structured data, move on. What's missing from this picture is the need to staple the type information to the data after you have type checked it, but that's essentially what this ZEP proposes, and what I disagree with.

Reply:
I think there seems to be a tension between making the format more robustly machine verifiable/machine parseable (embed schema URL) and making the format more readily human readable and human writable (use only short identifier).

If attributes have to be (in practice) identified by a URL, then it becomes challenging for humans to write the format except by copy and paste.

I can see the merits of both sides, though personally I am inclined towards human readable/writable, similar to HTML and CSS, where short identifiers are used.

I would be inclined to just say that the convention is identified by the name of the attribute itself, and there is no separate "zarr_connections" attribute, and it generally becomes easier to work with an edit, compared to having to keep both a dictionary of attributes and a list of conventions in sync.  It also avoids the possibility that two conventions would assign different meanings to the same attribute, which would prevent using both conventions at the same time.  If URLs are used to identify conventions, then the attribute name would itself be the URL, which would be awkward, though.

Reply:
@jbms, I like the idea of not having a separate `zarr_conventions`. Can you elaborate on your short identifier idea? Do you mean like the original idea in the ZEP; i.e. `units_v1` or something different? In this scenario, how would we avoid conflicts? If we make the key more elaborate, then the identifier is going to be not short.

Or are you thinking more of a hierarchical way to define the attributes? 

i.e. 

```json
{
  "units":
  {
    "length": "meter",
    "_convention": "< some-identifier >"
  }
}
```

```json
{
  "stats":
  {
    "std": 42,
    "mean": 100,
    "_convention": "< some-identifier >"
  }
}
```

Reply:
Yes, I mean something like `units_v1` --- a nested `_contention` seems worse than just using the convention as the property identifier itself.

However, I think we certainly do need a way to distinguish arbitrary non-standard metadata (which will presumably be used very widely) from standardized metadata properties that should be listed in some registry to ensure the identifier is unique.

I'm not sure exactly what sort of syntax makes sense --- possibly something like "std:units" or "zarr:units" or "$unit" or "@units".

However, I think the idea behind this zarr_conventions proposal is that you may already have a collection of datasets with various metadata, and software that consumes that metadata, and therefore do not want to change the representation of the metadata at all.  Instead you just want to tack on an additional property to indicate what metadata conventions are in use without having to modify the existing data or software.  Potentially these "legacy" conventions could still be handled as a separate property per-convention though --- e.g. you could set `"std:cf-conventions": true` but then there would be additional non-prefixed properties as defined by the convention.